crates/schema-inference/src/server.rs

```diff
@@ -231,38 +257,66 @@ async fn fragment_to_shape(
                             docs = docs + 1;
                         }
                     }
-                    Err(e) => return Err(InferenceError::from(e)),
+                    Some(Err(e)) => return Err(InferenceError::from(e)),
+                    None => break
                 }
             }
-            _ = owned_abort_channel.recv() => {
-                tracing::debug!(docs_processed = docs, "Aborting schma inference early!");
-                break
+            _ = abort_signal.changed() => {
+                if *abort_signal.borrow() {
+                    tracing::debug!("Aborting schma inference early!");
+                    break
+                }
             }
         }
     }
 
+    let bytes_read = doc_bytes_stream.decoder().bytes_read().try_into().unwrap();
+    tracing::debug!(
+        bytes_read = format!("{:.1} MB", bytes_read as f32 / 1_000_000f32),
+        docs_read = docs,
+        "Done reading fragment"
+    );
+
     match accumulator {
-        Some(accum) => Ok(Some((accum, docs))),
+        Some(accum) => Ok(Some(ShapeAndMeta {
+            shape: accum,
+            docs,
+            bytes: bytes_read,
+        })),
         None => Ok(None),
     }
 }
 
 async fn reduce_shape_stream(
-    stream: impl Stream<Item = Result<Option<(Shape, u64)>, InferenceError>>,
-) -> Result<Option<(Shape, u64)>, InferenceError> {
-    let mut accumulator: Option<(Shape, u64)> = None;
+    stream: impl Stream<Item = Result<Option<ShapeAndMeta>, InferenceError>>,
+) -> Result<Option<ShapeAndMeta>, InferenceError> {
+    let mut accumulator: Option<ShapeAndMeta> = None;
     tokio::pin!(stream);
 
-    while let Some(shape) = stream.next().await {
-        match shape {
-            Ok(Some((inferred_shape, docs_read))) => {
-                if let Some((accumulated_shape, docs_count)) = accumulator {
-                    accumulator = Some((
-                        shape::merge(accumulated_shape, inferred_shape),
-                        docs_count + docs_read,
-                    ))
+    while let Some(item) = stream.next().await {
```

Comment:
nit: this looks like a job for [`fold`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.fold)